### PR TITLE
Stop processing batch of fetched messages if partition is paused

### PR DESCRIPTION
**Pull request by tmattia**
Attempt to fix https://github.com/zendesk/ruby-kafka/issues/552

### CODE REVIEW

File: lib/kafka/consumer.rb [hunk 1]
```diff
@@ -229,7 +231,10 @@ def each_message(min_bytes: 1, max_bytes: 10485760, max_wait_time: 1, automatica
             @instrumenter.instrument("process_message.consumer", notification) do
               begin
                 yield message
-                @current_offsets[message.topic][message.partition] = message.offset
+
+                unless paused?(message.topic, message.partition)
+                  @current_offsets[message.topic][message.partition] = message.offset
+                end
```
Comment (benjaminoakes) [inline lib/kafka/consumer.rb]: I stepped through this issue as well and came to a very similar changeset.

File: lib/kafka/consumer.rb [hunk 2]
```diff
@@ -212,6 +212,8 @@ def each_message(min_bytes: 1, max_bytes: 10485760, max_wait_time: 1, automatica
             })
           end
           batch.messages.each do |message|
+            next if paused?(message.topic, message.partition)
```
Comment (dasch) [inline lib/kafka/consumer.rb]: This will effectively _skip_ all messages while a partition is paused, which is not the desired behavior.

**Comment by dasch** [comment]
I would like to make sure this use case is strictly necessary before adding more logic to the already complex processing code.

**Comment by tmattia** [comment]
Hey @dasch, thanks for taking the time to review this PR.

I'm not sure if I understand what you mean by skip.

From my understanding, the current behavior of the gem is unexpected because it continues to process messages from a partition even after the partition has been paused. The example given in the issue and the functional spec included in this PR reproduce the unexpected behavior.

According to the documentation for `Consumer#pause`, it should immediately stop processing messages from the paused partition. This is not currently happening because messages that have already been fetched (in a batch) are still processed even after the partition is paused.

The second issue with the current implementation is about where it resumes processing messages from a previously paused partition.

When `each_message` is called with `automatically_mark_as_processed: false`, it should resume from the last message that was not successfully marked as processed, allowing it a new chance at being successfully processed.

Does that make sense, or am I missing something? Is this against the behavior you would expect?

**Comment by dasch** [comment]
No, I think you're right to flag the current behavior as wrong.

File: lib/kafka/consumer.rb (hunk 1 above)
Comment (dasch) [reply lib/kafka/consumer.rb]: I think instead of doing this, we should break the loop for the entire batch. Let me take a look within the next few days.

**Comment by dasch** [comment]
I want to merge #303 first, then I'll fix this issue.

**Comment by tmattia** [comment]
@dasch thanks! Let me know if there's anything else you need, happy to help if you need more details or would like a different implementation.

**Pull request closed without merging**